File: src/inspect.rs
use qobuz_api::api::service::QobuzApiService;

use crate::{errors::AppError, search::fmt_duration, url::DownloadTarget};

const DIM: &str = "\x1b[2m";
const RESET: &str = "\x1b[0m";
const BOLD: &str = "\x1b[1m";
const RED: &str = "\x1b[31m";

pub fn run(api: &mut QobuzApiService, target: DownloadTarget, tsv: bool) -> Result<(), AppError> {
    let album_id = match target {
        DownloadTarget::Album(ref id) => id.clone(),
        DownloadTarget::Track(id) => {
            let track = api.get_track(id)?;
            track
                .album
                .as_ref()
                .and_then(|a| a.id.clone())
                .ok_or_else(|| AppError::Other("Track has no associated album".into()))?
        }
        _ => return Err(AppError::Other("inspect only supports album or track URLs".into())),
    };

    let album = api.get_album(&album_id, None)?;

    let artist = album.artist.as_ref().and_then(|a| a.name.as_deref()).unwrap_or("?");
    let title = album.title.as_deref().unwrap_or("?");
    let year = album
        .release_date_original
        .as_deref()
        .and_then(|d| d.split('-').next())
        .unwrap_or("?");
    let label = album.label.as_ref().and_then(|l| l.name.as_deref()).unwrap_or("?");
    let release_type = album
        .release_type
        .as_deref()
        .or(album.product_type.as_deref())
        .unwrap_or("album");
    let hires = album.hires_streamable == Some(true);
    let multi_disc = album.media_count.unwrap_or(1) > 1;

    let tracks = album.tracks.and_then(|t| t.items).unwrap_or_default();
    let locked_count = tracks.iter().filter(|t| t.streamable != Some(true)).count();

    if tsv {
        // Header row — album metadata
        println!(
            "album\t{}\t{}\t{}\t{}\t{}\t{}\t{}",
            esc(title),
            esc(artist),
            esc(year),
            esc(label),
            esc(release_type),
            if hires { "1" } else { "0" },
            locked_count,
        );
        // One row per track:
        // track_num \t disc \t title \t duration_secs \t streamable \t downloadable \t hires \t explicit
        for t in &tracks {
            let num = t.track_number.unwrap_or(0);
            let disc = t.media_number.unwrap_or(1);
            let track_title = t.title.as_deref().unwrap_or("?");
            let version = t.version.as_deref().unwrap_or("");
            let full_title = if version.is_empty() {
                track_title.to_string()
            } else {
                format!("{track_title} ({version})")
            };
            let dur = t.duration.unwrap_or(0);
            let streamable = if t.streamable == Some(true) { "1" } else { "0" };
            let downloadable = if t.downloadable == Some(true) { "1" } else { "0" };
            let track_hires = if t.hires == Some(true) { "1" } else { "0" };
            let explicit = if t.parental_warning == Some(true) { "1" } else { "0" };
            println!(
                "track\t{num}\t{disc}\t{}\t{dur}\t{streamable}\t{downloadable}\t{track_hires}\t{explicit}",
                esc(&full_title),
            );
        }
    } else {
        println!();
        println!("  {BOLD}{artist} — {title}{RESET}  ({year})  {label}  [{release_type}]{}",
            if hires { "  [Hi-Res]" } else { "" });
        println!("  {} track(s)", tracks.len());
        println!();

        if tracks.is_empty() {
            println!("  {DIM}(No track details returned){RESET}");
            println!();
            return Ok(());
        }

        for t in &tracks {
            let num = t.track_number.unwrap_or(0);
            let disc = t.media_number.unwrap_or(1);
            let track_title = t.title.as_deref().unwrap_or("?");
            let version = t.version.as_deref().unwrap_or("");
            let full_title = if version.is_empty() {
                track_title.to_string()
            } else {
                format!("{track_title} ({version})")
            };
            let dur = fmt_duration(t.duration.unwrap_or(0));
            let streamable = t.streamable == Some(true);
            let downloadable = t.downloadable == Some(true);
            let track_hires = t.hires == Some(true);
            let explicit = t.parental_warning == Some(true);

            let disc_prefix = if multi_disc { format!("{disc}-") } else { String::new() };

            let mut badges = String::new();
            if track_hires  { badges.push_str(" [hi-res]"); }
            if explicit     { badges.push_str(" [e]"); }
            if streamable && !downloadable { badges.push_str(" [stream-only]"); }

            if streamable {
                println!("  {disc_prefix}{num:02}. {full_title}  {DIM}{dur}{badges}{RESET}");
            } else {
                println!("  {DIM}{disc_prefix}{num:02}. {full_title}  {dur}{badges}  {RED}✗ locked{RESET}");
            }
        }

        println!();
        if locked_count == 0 {
            println!("  All {} track(s) available in this region.", tracks.len());
        } else {
            println!("  {RED}{locked_count} locked{RESET} / {} total — some tracks unavailable in this region.",
                tracks.len());
        }
        println!();
    }

    Ok(())
}

fn esc(s: &str) -> String {
    s.replace('\t', " ").replace('\n', " ").replace('\r', "")
}

File: src/inspect.cpp
#include "inspect.hh"
#include "search.hh"   // fmt_duration
#include "i18n.hh"

#include <cstdio>

#include <api/service.hh>
#include <core/models.hh>

namespace inspect {

kb::Result<void> run_album(kb::QobuzApiService &svc, const std::string &album_id) {
    auto res = svc.get_album(album_id);
    if (!res.ok()) return res.error();
    const auto &album = res.value();

    std::string title  = album.title.value_or("?");
    std::string artist = (album.artist && album.artist->name) ? *album.artist->name : "?";
    std::string label  = (album.label && album.label->name)   ? *album.label->name  : "?";
    std::string date   = album.release_date_original.value_or("?");
    std::string genre  = (album.genre && album.genre->name)   ? *album.genre->name  : "?";
    int total_dur      = album.duration.value_or(0);
    int tracks_cnt     = album.tracks_count.value_or(0);

    bool hi = album.hires.value_or(false) || album.hires_streamable.value_or(false);
    const char *qual = "?";
    if (hi) {
        int bd = album.maximum_bit_depth.value_or(0);
        double sr = album.maximum_sampling_rate.value_or(0.0);
        if (bd >= 24 && sr >= 192.0)     qual = "flac-ultra";
        else if (bd >= 24 && sr >= 96.0) qual = "flac-hi";
        else                              qual = "flac";
    } else {
        qual = "mp3/flac";
    }

    std::printf("%s: %s\n", i18n::t("field_title"),  title.c_str());
    std::printf("%s: %s\n", i18n::t("field_artist"), artist.c_str());
    std::printf("%s: %s\n", i18n::t("field_label"),  label.c_str());
    std::printf("%s: %s\n", i18n::t("field_date"),   date.c_str());
    std::printf("%s: %s\n", i18n::t("field_genre"),  genre.c_str());
    std::printf("%s: %s\n", i18n::t("field_quality"), qual);
    std::printf("%s: %s (%d %s)\n", i18n::t("field_duration"),
                search::fmt_duration(total_dur).c_str(),
                tracks_cnt, i18n::t("tracks_suffix"));

    if (!album.tracks) {
        std::printf("\n%s\n", i18n::t("no_tracks"));
        return {};
    }

    auto &tr_res = *album.tracks;
    auto &items  = tr_res.items;
    if (!items || items->empty()) {
        std::printf("\n%s\n", i18n::t("no_tracks"));
        return {};
    }

    std::printf("\n");
    int media = -1;
    for (const auto &tp : *items) {
        if (!tp) continue;
        const auto &t = *tp;

        int disc = t.media_number.value_or(1);
        if (disc != media) {
            media = disc;
            std::printf("%s %d\n", i18n::t("disc"), disc);
        }

        int    num    = t.track_number.value_or(0);
        int    dur    = t.duration.value_or(0);
        bool   dl_ok  = t.downloadable.value_or(false);
        bool   hi_t   = t.hires.value_or(false);
        std::string ttitle = t.title.value_or("?");
        std::string tver   = t.version.value_or("");

        const char *avail = dl_ok ? "[DL]" : "    ";
        const char *hires = hi_t  ? " [Hi-Res]" : "";

        if (tver.empty())
            std::printf("  %2d. %-40s  %s  %s%s\n",
                num, ttitle.c_str(),
                search::fmt_duration(dur).c_str(), avail, hires);
        else
            std::printf("  %2d. %-40s  %s  %s%s\n",
                num, (ttitle + " (" + tver + ")").c_str(),
                search::fmt_duration(dur).c_str(), avail, hires);
    }
    return {};
}

kb::Result<void> run_track(kb::QobuzApiService &svc, int track_id) {
    auto res = svc.get_track(track_id);
    if (!res.ok()) return res.error();
    const auto &t = res.value();

    std::string title  = t.title.value_or("?");
    std::string tver   = t.version.value_or("");
    std::string artist = (t.performer && t.performer->name) ? *t.performer->name : "?";
    std::string album_t = (t.album && t.album->title) ? *t.album->title : "?";
    std::string date   = t.release_date_original.value_or("?");
    int         dur    = t.duration.value_or(0);
    bool        dl_ok  = t.downloadable.value_or(false);
    bool        hi_t   = t.hires.value_or(false);

    std::printf("%s: %s%s\n", i18n::t("field_title"),
                title.c_str(), tver.empty() ? "" : (" (" + tver + ")").c_str());
    std::printf("%s: %s\n", i18n::t("field_artist"),   artist.c_str());
    std::printf("%s: %s\n", i18n::t("field_album"),    album_t.c_str());
    std::printf("%s: %s\n", i18n::t("field_date"),     date.c_str());
    std::printf("%s: %s\n", i18n::t("field_duration"), search::fmt_duration(dur).c_str());
    std::printf("%s: %s\n", i18n::t("field_hires"),    hi_t  ? i18n::t("yes") : i18n::t("no"));
    std::printf("%s: %s\n", i18n::t("field_download"), dl_ok ? i18n::t("yes") : i18n::t("no"));
    return {};
}

static std::string esc_tsv(const std::string &s) {
    std::string r;
    for (char c : s) { if (c == '\t' || c == '\n' || c == '\r') r += ' '; else r += c; }
    return r;
}

kb::Result<void> run_album_tsv(kb::QobuzApiService &svc, const std::string &album_id) {
    auto res = svc.get_album(album_id);
    if (!res.ok()) return res.error();
    const auto &album = res.value();

    std::string title  = esc_tsv(album.title.value_or(""));
    std::string artist = (album.artist && album.artist->name) ? esc_tsv(*album.artist->name) : "";
    std::string year   = album.release_date_original.value_or("").substr(0, 4);
    std::string label  = (album.label && album.label->name) ? esc_tsv(*album.label->name) : "";
    std::string rtype  = esc_tsv(album.release_type.value_or(""));
    int hires_flag     = (album.hires.value_or(false) || album.hires_streamable.value_or(false)) ? 1 : 0;

    // Count locked (not downloadable) tracks
    int locked_count = 0;
    if (album.tracks && album.tracks->items) {
        for (auto &tp : *album.tracks->items) {
            if (tp && !tp->downloadable.value_or(true)) ++locked_count;
        }
    }

    std::printf("album\t%s\t%s\t%s\t%s\t%s\t%d\t%d\n",
        title.c_str(), artist.c_str(), year.c_str(),
        label.c_str(), rtype.c_str(), hires_flag, locked_count);

    if (!album.tracks || !album.tracks->items) return {};
    for (auto &tp : *album.tracks->items) {
        if (!tp) continue;
        const auto &t = *tp;
        std::string ttitle = esc_tsv(t.title.value_or(""));
        if (t.version) ttitle += " (" + esc_tsv(*t.version) + ")";
        std::printf("track\t%d\t%d\t%s\t%d\t%d\t%d\t%d\t%d\n",
            t.track_number.value_or(0),
            t.media_number.value_or(1),
            ttitle.c_str(),
            t.duration.value_or(0),
            t.streamable.value_or(false) ? 1 : 0,
            t.downloadable.value_or(false) ? 1 : 0,
            t.hires.value_or(false) ? 1 : 0,
            (t.parental_warning == std::optional<bool>{true}) ? 1 : 0);
    }
    return {};
}

} // namespace inspect
